🦺 Update implementation and docs for `RequestValidationError` in Pydantic v2

tamird opened the pull request:
Change the signature of `ValidationException.errors` to return
`Sequence[pydantic_core.ErrorDetails]` in Pydantic V2.

Update the documentation to explain that `RequestValidationError` isn't
literally a subclass since Pydantic V2.

Closes (**UPD YuriiMotov:** actually not) #10424.
Closes #11176.

tamird committed:
ValidationException.errors() are ErrorDetails

Change the signature of `ValidationException.errors` to return
`Sequence[pydantic_core.ErrorDetails]` in Pydantic V2.

Add type annotations in tests and locally run mypy to ensure it fails if
the annotations changes to e.g. `Sequence[str]`. Unfortunately running
mypy over tests isn't currently part of CI.

tiangolo commented:
📝 Docs preview for commit aa59b77ac27f4ef842be10d8e7be9957f6478c43 at: https://69746e9e.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit 25da9fdc0af2f3249acd7b32fee94a59086bbc78 at: https://a9ca36dc.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit 45a560a69c4a02da61678a82e77ac427dc88f177 at: https://787fcba4.fastapitiangolo.pages.dev

tamird committed:
Remove unused suppressions

tiangolo commented:
📝 Docs preview for commit 338a33e2b9eecb9138d29fa470634b279c5956d0 at: https://5e014aa1.fastapitiangolo.pages.dev

tamird commented:
@tiangolo this is green now. I believe it is technically a breaking change.

tiangolo commented:
📝 Docs preview for commit 796e817091c08dbcfc530512965f38873ea0ae86 at: https://7516403b.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit de154678a1f2fd8789166c50ba4ed95427861e31 at: https://b04b1a0e.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit 30788f5cffae8fd2c9e00258b4b194d56c20b011 at: https://18f55b1d.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit f1b3e2ff0ef65d1da795987b2bfc8051ea1b2479 at: https://d5332d85.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit 11aca70a419392f1b37b1e6e65070be7d0f8fd93 at: https://a014911b.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit edca75902fc452fa9090a169feb83a49f99186b8 at: https://9fa0665c.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit 025ef5fd714f8f0268ca9dd81a281947393bcb6f at: https://3e363dc1.fastapitiangolo.pages.dev

tiangolo commented:
📝 Docs preview for commit b91428de008aa0c79729006068021b84d1575feb at: https://b683ce8c.fastapitiangolo.pages.dev

tamird commented:
@tiangolo ping - would you mind having a look please?

tamird commented:
@alejsdev could you please have a look? What is the process for getting a review?

tamird commented:
@tiangolo ping

bertomaniac commented:
I think this PR will actually fix the issue I described here:
https://github.com/fastapi/fastapi/discussions/11923

The extra `input` and `ctx` fields that come out of pydantic v2 `ValidationError.errors()` is causing problems when serializing that data back in `fastapi`.

tamird commented:
3 months and no review :(

bertomaniac commented:
Dang, I just checked out your PR and tried it for my test. It doesn't fix it because I think this code block still needs to strip out `ctx` and `input` from pydantic v2s `ValidationError.errors()` call:
```
            except ValidationError as exc:
                errors: List[ErrorDetails] = [
                    {**err, "loc": loc + err.get("loc", ())}  # type: ignore [typeddict-unknown-key]
                    for err in exc.errors(include_url=False)
                ]
                return None, errors
```

perhaps
```
for err in exc.errors(include_url=False, include_input=False, include_context=False)
```

See here in pydantic docs: https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.ValidationError.errors

estebanx64 commented:
Hi @tamird 👋

We'll be reviewing this PR soon. 🤓

Thank you for your interest and time in helping us with your contribution 🙇

ecly commented on docs/en/docs/tutorial/handling-errors.md:
```diff
@@ -163,7 +163,7 @@ path -> item_id
 !!! warning
```
@tamird due to the lack of subclassing, the text version of the validation above this is also not the expected output. See: https://github.com/fastapi/fastapi/discussions/12125

Probably makes sense to update this text version as well.

tamird replied:
Sorry, can you suggest a patch? It has been months since I submitted this PR.

tamird replied:
Updating the text is not the right solution. See https://github.com/fastapi/fastapi/pull/12972.

tamird commented:
I have rebased this to resolve the merge conflicts.

tamird commented:
@estebanx64 ping.

tiangolo commented:
@tamird thanks for the interest. There's no need to ping or contact individual maintainers, we already have this PR in the project and we'll review it eventually.

If you would like to help accelerate things, there are many ways to contribute, for example helping others with questions: https://fastapi.tiangolo.com/help-fastapi/#help-others-with-questions-in-github

YuriiMotov commented:
@tamird, please refrain from rebasing branch, it creates lots of comments that makes it harder to review PR.

Also, please update the description: currently changes not just update docs, but change code

YuriiMotov reviewed (commented):
This PR modifies `ValidationException` and derived classes (`RequestValidationError` and `ResponseValidationError`) so that their `.errors()` method would return `Sequence["ErrorDetails"]` instead of `Sequence[Any]`.

Where `ErrorDetails` is a `TypedDict` imported from Pydantic (`pydantic.error_wrappers.ErrorDict` for V1 and `pydantic_core.ErrorDetails` for V2).
```python
class ErrorDetails(_TypedDict):
    type: str
    loc: tuple[int | str, ...]
    msg: str
    input: _Any
    ctx: _NotRequired[dict[str, _Any]]
    url: _NotRequired[str]
```

So, I guess the intention is to provide type hints in case user write something like:

```python
from fastapi import Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse

async def request_validation_exception_handler(
    request: Request, exc: RequestValidationError
) -> JSONResponse:
    errors = exc.errors()  # Here I expect to ge list of `ErrorDetails`
    ...
```

But in VSCode I can still see that elements of `errors` are of type `Any`:
<img width="516" height="168" alt="image" src="https://github.com/user-attachments/assets/f6ccd00e-f781-400a-80bd-d665e19702f8" />

@tamird, could you please take a look?

tamird commented:
> But in VSCode I can still see that elements of `errors` are of type `Any`: <img alt="image" width="516" height="168" src="https://private-user-images.githubusercontent.com/109919500/469212082-f6ccd00e-f781-400a-80bd-d665e19702f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMxOTQzODMsIm5iZiI6MTc1MzE5NDA4MywicGF0aCI6Ii8xMDk5MTk1MDAvNDY5MjEyMDgyLWY2Y2NkMDBlLWY3ODEtNDAwYS04MGJkLWQ2NjVlMTk3MDJmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMlQxNDIxMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMmU1NTc2NzhlODk5ODI5NzczNjBkOGI1OWU1MzMzYzZiZDNkMzM2YmIxYzFkZWQ0OTVkMWZhNmMzYzk1MmMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Ueh_9a5_i6d2ImlxfNbDTO2_oCDlnfUO1i21zptVQ0g">
> 
> @tamird, could you please take a look?

I do not see the token "Any` in your screenshot.

YuriiMotov commented:
> I'm not sure how your local environment is set up, and I'm not sure exactly what code you're looking at.

I looked at the code from [this comment](https://github.com/fastapi/fastapi/pull/11542#pullrequestreview-3042531180).

And I use fresh virtual environment with the latest versions of packages installed using `pip install -r requirements.txt`.

I don't think this is a problem of my environment set up. If I create my own function with the same return type annotation, it works well
<img width="335" height="108" alt="image" src="https://github.com/user-attachments/assets/02d5e6e9-2376-49fb-8a71-d4720e9af114" />

And if I try the same with:
```python
from fastapi.exceptions import ValidationException

exc = ValidationException([])

b = exc.errors()[0]
```
It gives my `Any`

Full code example that you or anybody else can try:
```python
from collections.abc import Sequence

from fastapi.exceptions import ValidationException
from pydantic_core import ErrorDetails


# =============================================
def my_errors() -> Sequence["ErrorDetails"]:
    pass

a = my_errors()[0]  # ErrorDetails


# =============================================
exc = ValidationException([])

b = exc.errors()[0]  # Any

```


> Rather than post a screenshot, perhaps you could add a type annotation in the relevant place?

> A test change would be a robust way to ensure that the type annotations are working as intended.

Sorry, I didn't get the above messages. Could you please clarify what you mean?

tamird committed:
Remove Pydantic V1 TODOs

These tests are all marked `@needs_pydanticv2` so they don't run on
Pydantic V1.

tamird committed:
Add type annotations to pydanic model tests

This will allow mypy to check the return type of
`ValidationException.errors()` in a later commit.

tamird commented:
@YuriiMotov I have added type annotations in the relevant tests and confirmed locally that changing the annotations to `Sequence[str]` does cause mypy to complain (see updated commit message).

tamird commented:
It looks like the `if TYPE_CHECKING` guards are interfering with the type checker, at least in the IDE.

Let me play with it some more.
